### ISSUE TITLE
Server gets stuck, apparently due to broken Cargo metadata

### ISSUE THREAD

**Issue by djc** [rust-analyzer/rust-analyzer#6589]
So I've run an issue a few times now over the last few weeks where my cargo check commands in the terminal get stuck for a very long time (sometimes minutes) blocking on "waiting for file lock on package cache". I just had this happen to me again, and decided to dig in a little. I now think this might be happening when RA tries to read my `Cargo.toml` file just as I'm editing it. This gets it into an error state, like this:

```
[ERROR rust_analyzer::reload] failed to load workspace: Failed to read Cargo metadata from Cargo.toml file /Users/djc/src/ids-rs/Cargo.toml, cargo 1.47.0 (f3c7e066a 2020-08-28): Failed to run `cargo metadata --manifest-path /Users/djc/src/ids-rs/Cargo.toml`: Error during execution of `cargo metadata`: error: failed to parse manifest at `/Users/djc/src/ids-rs/nominl/Cargo.toml`

Caused by:
  could not parse input as TOML

Caused by:
  newline in string found at line 24 column 96

[ERROR flycheck] Flycheck failed to run the following command: "cargo" "check" "--workspace" "--message-format=json" "--manifest-path" "/Users/djc/src/ids-rs/Cargo.toml" "--all-targets" "--all-features"
[ERROR rust_analyzer::main_loop] cargo check failed: Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(ExitStatus(25856)))
```

After this, it looks like it eventually recover, but it's taking very long to do so (maybe some timeout?) -- and all the while it appears to be holding the package cache file lock.

**Comment by mathemaphysics** [comment]
Same thing happening to me. Doesn't matter what I do it just blocks saying that it's waiting for cargo metadata or cargo check. Closing all files doesn't do anything.

**Comment by bluretrece** [comment]
@mathemaphysics I'm currently dealing with the same issue. Did you were able to solve it?

**Comment by mathemaphysics** [comment]
@bluretrece It turns out that there's a command that gets blocked by a request for user info. It's a cargo subcommand. I'll look it up in my history and post.

**Comment by bluretrece** [comment]
@mathemaphysics please, I'd appreciate it a lot! rust-analyzer is still getting stuck in all of my projects. I opened an issue describing the bug last night, but still got no responses whatsoever.

**Comment by chrisguida** [comment]
I am also getting this error, disabling, uninstalling, reinstalling, etc do not help

**Comment by clarammdantas** [comment]
I am also getting this error and also tried to disabling, uninstalling, reinstalling, and all of that. It didn't solve the problem

**Comment by Veykril** [comment]
We need a reproduction here to be able to work on this.

**Comment by clarammdantas** [comment]
I finally found a solution that worked for me, and what did the trick was to open the project with a root as the rust project I'm working on. Before I opened two directories up because the rust project is part of a large mono repo project.

**Comment by flipbit03** [comment]
To whoever is losing hours of debugging time puzzled about this, because I struggled too and saw posts and some threads which spanned for months were people where lost, and I myself lots a couple of useful hours on this stupid thing.

I was having the same problem with a project that uses a Devcontainer. So you have a Dockerfile that defines your environment and you develop from inside it.

I decided to base my Devcontainer off `rust:1.XX-bullseye` instead of the microsoft sanctioned ready-to-ingest pill, and whenever I'd save a .rs file, the message would pop up:

```cargo check failed: Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(unix_wait_status(256)))```

Turns out the only problem is that `cargo clippy` was not available. All you need to do, to fix this, is add `cargo clippy` via this command:

    rustup component add clippy

This is not a problem technically related to rust-analyzer. But from an user experience standpoint, rust-analyzer expects cargo clippy to be available, and it could display an informative message or hint together with the cargo's error mesage. Something along the lines of :

```
cargo check failed: Cargo watcher failed, the command produced
no valid metadata (exit code: ExitStatus(unix_wait_status(256))).
Do you have cargo clippy installed?
```

This would save unadvised users hours of grief.

Thanks in advance, and hope this answer can be useful to someone.

**Comment by camelid** [comment]
I ran into a similar error while working on rustc itself. I tried deleting the build cache, restarting the r-a server, and restarting VS Code, without success. Formatting is working, but this error message keeps appearing:

```
cargo check failed: Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(unix_wait_status(32512)))
```

and no errors are being displayed in the UI (even though my code is incorrect). `x.py check` works fine.

I have clippy installed for all toolchains via rustfmt (though I guess not the custom toolchain used by `x.py`).

**Comment by Veykril** [comment]
Can you try the nightly version of r-a? That one should show you the actual error message

**Comment by camelid** [comment]
Aha, the issue is that it's trying to use `python` but I only have `python3`:

```
cargo check failed: Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(unix_wait_status(32512))) Cargo's stderr output: env: python: No such file or directory
```

I think this is an issue on x.py's end since its shebang requests `python` but that may not exist on newer computers. Indeed, if I run `./x.py check`, I get `env: python: No such file or directory`. I noticed this earlier but didn't connect the dots to the r-a issue. I'll ask T-compiler about this. Thanks!

**Comment by lattice0** [comment]
neither `rustup component add clippy` or `rustup update` fixed for me

**Comment by Veykril** [comment]
Can you also try the nightly version of r-a? That one should show you the actual error message.

There isn't much we can do here without a reproduction otherwise. I'll go ahead and close this issue as there is not much we can do, and we improved the error output for most cases.
If you run into this and the extra error output from current nightly (or next week's stable) is unhelpful you can open a new issue with sufficient info for us.

**Comment by lattice0** [comment]
On pre-release I got

```
cargo check failed: Cargo watcher failed, the command produced no valid metadata (exit code: ExitStatus(unix_wait_status(25856)))
Cargo's stderr output:
error: no such subcommand: ``

	Did you mean `b`?
```

and on VSCode output:
`
[ERROR flycheck] Flycheck failed to run the following command: "cargo" "" "--workspace" "--message-format=json" "--manifest-path" "/home/dev/orwell/Cargo.toml" "--all-targets"`

Looks like on of the subcommands is `""`

**Comment by Veykril** [comment]
Looks like you set your `rust-analyzer.checkOnSave.command` to an empty string
